riptable/rt_str.py

```diff
@@ -707,12 +723,88 @@ def endswith(self, str2):
         FastArray([True, False, False])
         '''
         if not isinstance(str2, FAString):
+            if str2 == '':
+                return ones(self.n_elements, dtype=bool)
+
             str2 = self.possibly_convert_tostr(str2)
             if len(str2) != 1:
                 return TypeError(f"A single string must be passed for str2 not {str2!r}")
             str2 = FAString(str2)
 
-        return self._apply_func(self.nb_endswith, self.nb_endswith, str2, dtype=np.bool)
+        return self._apply_func(self.nb_endswith, self.nb_endswith, str2, dtype=np.bool,
+                                filtered_fill_value=False)
+
+    def regexpb(self, str2):
+        '''
+        Return a Boolean array where the value is set True if the string contains str2.
+        str2 may be a normal string or a regular expression.
+        Applies re.search on each element with str2 as the pattern.
+
+        Examples
+        --------
+        >>> FAString(['abab','ababa','abababb']).regexpb('ab$')
+        FastArray([True, False, False])
+        '''
+        if not isinstance(str2, bytes):
+            str2 = bytes(str2, 'utf-8')
+        str2 = re.compile(str2)
+        vmatch = np.vectorize(lambda x: bool(str2.search(x)))
+        bools = vmatch(self.backtostring)
+        if self._ikey is not None:
+            bools = bools[self._ikey] & (self._ikey >= 0)
+        return bools
+
+    @staticmethod
+    @nb.njit(cache=False)

```

Comment:
Is there any way to re-work your algorithm here to support parallelism, e.g. so you can set ``parallel=True`` and use ``nb.prange()`` somewhere below? One goal of riptable is for all operations to be multithreaded/parallelized by default, with the exceptions being cases where it would be extremely difficult or impossible to implement a parallelized algorithm.

Reply:
> Why are you setting `cache=False` here?

I am actually not sure. I just copied it from the other algos above. I'll give the parellelization a go.

Reply:
> Is there any way to re-work your algorithm here to support parallelism, e.g. so you can set `parallel=True` and use `nb.prange()` somewhere below? One goal of riptable is for all operations to be multithreaded/parallelized by default, with the exceptions being cases where it would be extremely difficult or impossible to implement a parallelized algorithm.

I tried putting in prange in the outer loop but doesn't appear to have made a difference. One thing to note is that my implementation relies in strlen which take a similar amount of time. I was thinking this could be made a cached_property to cut the substr time in half. Also, strlen is not parallelized. I have come up with a potentially better implementation that iterates over each string from back to front. This assumes that the string lengths tend to be closer to the max length than to 0.

Reply:
How big of an array did you test/benchmark with? You'd only see the benefits of ``nb.prange`` for larger arrays -- "large" there depends on the algorithm, machine, numba version, etc. but 1M elements should be plenty to tell whether it's working or not.

If you're testing with a large-enough array and still not seeing any benefits, maybe _numba_ is having trouble understanding how to parallelize the loop. You can enable parallel / codegen diagnostics and try running the code to get more info from numba about this: https://numba.pydata.org/numba-doc/dev/user/parallel.html#diagnostics

Your idea about caching the result of ``strlen`` is interesting. We won't be able to make it a ``@cached_property`` until we drop support for Python 3.7 though, or we'll have to come up with some workaround for that version.

Reply:
> Why are you setting `cache=False` here?

Thanks for the tip on the diagnostics. The substr implementation now works in parallel. I was worried updating `max_chars` at each iteration might not work in parallel but I see from the docs that it can handle this. I've also added a parallel version of `strlen` and applied `cached_property` where available.